src/lib/Embed.hs

```diff
@@ -292,24 +285,34 @@ ieq :: MonadEmbed m => Atom -> Atom -> m Atom
 ieq x@(Con (Lit _)) y@(Con (Lit _)) = return $ applyIntCmpOp (==) x y
 ieq x y = emitOp $ ScalarBinOp (ICmp Equal) x y
 
+fromPair :: MonadEmbed m => Atom -> m (Atom, Atom)
+fromPair pair = do
+  scope <- getScope
+  let pair' = reduceAtom scope pair
+  return (getProjection [0] pair', getProjection [1] pair')
+
 getFst :: MonadEmbed m => Atom -> m Atom
-getFst (PairVal x _) = return x
-getFst p = emitOp $ Fst p
+getFst p = fst <$> fromPair p
 
 getSnd :: MonadEmbed m => Atom -> m Atom
-getSnd (PairVal _ y) = return y
-getSnd p = emitOp $ Snd p
+getSnd p = snd <$> fromPair p
 
 getFstRef :: MonadEmbed m => Atom -> m Atom
 getFstRef r = emitOp $ FstRef r
 
 getSndRef :: MonadEmbed m => Atom -> m Atom
 getSndRef r = emitOp $ SndRef r
 
+-- XXX: getUnpacked must reduce its argument to enforce the invariant that
+-- ProjectElt atoms are always fully reduced (to avoid type errors between two
+-- equivalent types spelled differently).
```

Comment:
Oh, that would be neat! And then maybe this part could be simplified, because then there would be only one representation of any given atom. (I think there are at least a few places where we `emit` atoms and expect to get `Var`s out, though, so we'd have to fix those.)

Reply:
Yeah this for sure doesn't hold right now, but it might be an interesting thing to consider. I don't think we depend on those `Var`s in too many palces.